File: trunk/src/pypann/Core/Net.py
#
# NetCache
#
class NetCache:
    def __init__(self):
        self._ok = False
        self.layers = []

    def is_ok(self):
        return self._ok

    def invalidate(self):
        self._ok = False

    def fix(self):
        self._ok = True

#
# Net
#
import multiprocessing
from Link import Link
from Queue import Queue
from threading import Thread

class Net:
    def __init__(self):
        self.input_neurons = []
        self.worker_threads_count = multiprocessing.cpu_count()
        self._cache = NetCache()

    def add_input_neuron(self, n):
        self.input_neurons.append(n)
        self._cache.invalidate()

    def remove_neuron(self, n):
        for link in n._links_in:  self.disconnect(link.to(), n)
        for link in n._links_out: self.disconnect(n, link.to())

        try:
            self.input_neurons.remove(n)
        except ValueError:
            pass

        del n

        self._cache.invalidate()

    def connect(self, n_from, n_to, w, latency = 1):
        n_from._links_out.append(Link(n_to, w, latency))
        n_to._links_in.append(Link(n_from, w, latency))
        self._cache.invalidate()

    def disconnect(self, n_from, n_to):
        n_from._links_out = [link for link in n_from._links_out if link.to() != n_to]
        n_to._links_in = [link for link in n_to._links_in if link.to() != n_from]
        self._cache.invalidate()

    def set_input(self, values):
        for i in range(len(self.input_neurons)):
            self.input_neurons[i].input = values[i]

    def get_output(self):
        if not self._cache.is_ok():
            self._update_cache()

        if len(self._cache.layers):
            return map(lambda n: n.output, self._cache.layers[-1])
        else:
            return []

    #dir - True - forward run, False - backward
    def run(self, runner, dir = True):
        if not self._cache.is_ok():
            self._update_cache()

        def worker():
            while True:
                neuron = q.get()
                runner(neuron)
                q.task_done()

        q = Queue()
        for i in range(self.worker_threads_count):
            t = Thread(target=worker)
            t.setDaemon(True)
            t.start()

        for l in self._cache.layers:
            for n in l:
                q.put(n)
            q.join()


        pass #TODO join all

    def _update_cache(self):
        self._cache.layers = []
        hops = dict(map(lambda x: (x, 1), self.input_neurons))

        layer = set(self.input_neurons)
        while len(layer):
            self._cache.layers.append(layer)
            next_layer = set()
            for n in layer:
                for link in n._links_out:
                    if not link.to() in hops:
                        hops[link.to()] = hops[n] + link.latency()
                    if hops[link.to()] == hops[n] + 1:
                        next_layer.add(link.to())
                layer = next_layer

        self._cache.fix()

#
# Testing
#
from Weight import Weight
from Neuron import *

def test_Net():
    print "Testing Net..."

    N = Net()
    print N.get_output()

    n = {}
    for i in range(4):
        n[i] = Neuron()

    def runner(N):
        pass

    def print_cache(N):
        print "Cache:"
        N.run(runner)
        for l in N._cache.layers:
            print [x for x in range(len(n)) for neu in l if n[x] == neu]

    N.add_input_neuron(n[0])
    N.connect(n[0], n[1], Weight(3.0))
    N.connect(n[0], n[2], Weight())
    N.connect(n[1], n[3], Weight())
    N.connect(n[2], n[3], Weight())
    print_cache(N)
    print N.get_output()

    N.remove_neuron(n[3])
    print_cache(N)

#
# Main
#
if __name__ == "__main__":
    test_Net()

File: trunk/src/pann/Core/Net.cpp
//Net.cpp

#include <boost/foreach.hpp>

#include "Runner.h"
#include "Neuron.h"

#include "Net.h"

using namespace std;

namespace pann
{
    Net::Net()
    {
        set_work_threads_count(0);
    } //Net

    Net::~Net()
    {
    } //~Net

    void
    Net::add_input_neuron(NeuronPtr neuron)
    {
        cache.touch();

        input_neurons.push_back(neuron);
    } //add_input_neuron

    void
    Net::remove_neuron(NeuronPtr neuron)
    {
        cache.touch();

        //Delete all connections to/from current neuron
        BOOST_FOREACH(const Link& link, neuron->input_links)
            link.get_to()->remove_out_connection(neuron);

        BOOST_FOREACH(const Link& link, neuron->output_links)
            link.get_to()->remove_in_connection(neuron);

        //Remove neuron from registers
        input_neurons.remove(neuron);

        Debug()<<neuron.use_count()<<'\n';
    } //remove_neuron

    WeightPtr
    Net::add_connection(NeuronPtr from, NeuronPtr to, WeightPtr weight)
    {
        cache.touch();

        if(!weight)
            weight.reset(new Weight());

        from->add_out_connection(to, weight);
        to->add_in_connection(from, weight);

        return weight;
    } //add_connection

    void
    Net::remove_connection(NeuronPtr from, NeuronPtr to)
    {
        cache.touch();

        /* Short unsignedroduction:
         * Neuron_from               Neuron_to
         *  Link                      Link
         *    to =>Neuron2              to =>Neuron1
         *    w  =>weight1              w  =>weight1
         *    dir=>out                  dir=>in
         */

        from->remove_out_connection(to);
        to->remove_in_connection(from);
    } //remove_connection

    void
    Net::set_input(const valarray<Float>& input)
    {
        if(input.size() < input_neurons.size())
            throw SizeMismatch()<<"set_input(): Supplied input size is smaller "
                                     "then number of input neurons\n";

        if(input.size() > input_neurons.size())
            Warning()<<"set_input(): Input size is bigger then input neurons count. "
                             "Check getInputMap() output\n";

        unsigned i = 0;
        BOOST_FOREACH( const NeuronPtr& n, input_neurons)
           n->input = input[i++];
    } //set_input

    void
    Net::get_output(valarray<Float>& output) const
    {
        if(!cache.is_ok())
            regenerate_cache();

        vector<NeuronPtr>& last_layer = *(cache.layers.end() - 2);
        unsigned output_size = last_layer.size();

        output.resize(output_size);
        for(unsigned i = 0; i < output_size; ++i)
            output[i] = last_layer[i]->output;
    } //get_output

    void
    Net::run(RunnerPtr runner, RunDirection direction)
    {
        if( !cache.is_ok() )
            regenerate_cache();

        boost::thread_group thread_pool;
        boost::barrier barrier(work_threads_count);

        //We must give parameters by pointer, because boost will copy all arguments to thread
        for(unsigned thread = 0; thread < work_threads_count; ++thread)
            thread_pool.add_thread( new boost::thread(Net::thread_base, direction, runner, this, thread, &barrier) );

        //wait for threads to finish
        thread_pool.join_all();
    } //run

    const NetCache&
    Net::get_cache() const
    {
        if(!cache.is_ok())
            regenerate_cache();

        return cache;
    } //get_cache

    unsigned
    Net::get_work_threads_count() const
    {
        return work_threads_count;
    } //get_work_threads_count

    void Net::set_work_threads_count(unsigned count)
    {
        if(count <= 0)
            work_threads_count = boost::thread::hardware_concurrency();
        else
            work_threads_count = count;

    } //set_work_threads_count

    void
    Net::format_front(list<NeuronPtr>& raw) const
    {
        cache.layers.push_back(vector<NeuronPtr>());

        if(raw.size() > 0)
        {
            raw.sort();
            raw.unique();

            for(list<NeuronPtr>::iterator it = raw.begin(); it != raw.end(); ++it)
                cache.layers.back().push_back(*it);
        }
    } //format_front

    void
    Net::regenerate_cache() const
    {
        //TODO: if user set's latency > 1 for regular links - shit may happen
        //TODO: Latency - is very dangerous tool. It's not difficult to understand, 
        //TODO: why this feature is still not implemented in Neuron/Net interface :)
        cache.flush();

        //Here we will place neuron's IDs that will become front, with duplicates
        list<NeuronPtr> raw_front;

        /*
         * Function operates with "hops" attribute of every Neuron
         * to build adequate cache.
         */
        map<NeuronPtr, unsigned> hops;

        //Put inputNeurons to front
        BOOST_FOREACH( const NeuronPtr& n, input_neurons )
        {
            raw_front.push_back(n);
            hops[n] = 1;
        }
        format_front(raw_front);

        /*
         * Cache looks like this:
         * At this point cache.layers[0] is the only record and it contains input neurons
         */
        do {
            /*
             * We have to prepare next front
             * At first iteration, if you remember, vector<NeuronPtr> rawFront contains unique inputNeurons
             */
            unsigned nCount = raw_front.size();
            Debug()<<nCount<<"\n";
            for(unsigned i = 0; i < nCount; ++i)
            {
                //pop_front emulation
                NeuronPtr current_neuron = raw_front.front();
                raw_front.pop_front();

                //ok, we've got cur_neuron. We will iterate through his Out links
                //and push_back their opposite sides to rawFront
                BOOST_FOREACH( const Link& link, current_neuron->output_links )
                {
                    const NeuronPtr& to_neuron = link.get_to();
                    /*
                     * Short comment. Consider following topology (hops are in brackets):
                     *   +(0)    <-- output          +(0)             +(3)    <-- front(3)
                     *  / \                         / \              / \
                     * +(0)+(0)         ======>    +(0)+(0)  ====>  +(2)+(2)  <-- front(2)
                     *  \ /           iteration1    \ /        N     \ /
                     *   +(0)  <-- input             +(1)             +(1)    <-- front on iteration1
                     *
                     *   For current neuron (C) and other neuron (T), hops might be:
                     *   T = 0  - T-neuron is fresh. We will set T=C+1
                     *   T = N, N < C - it is recurrent link. Don't touch T and don't place it to rawFront
                     *   T = C  - it is recursive topology(recursion in same layer)
                     *   T = C, but t=c  - it is recurrent link over 1 neuron
                     *   T = C + 1 - T already handled. We should add T to rawFront
                     *   T > C + 1 - latency to T is > 1. Don't touch T.
                     */

                    //Assume that when cache becomes coherent, all neuron[hops] vars become zero
                    if(hops[to_neuron] == 0)
                        hops[to_neuron] = hops[current_neuron] + link.get_latency();

                    if(hops[to_neuron] == hops[current_neuron] + 1)
                        raw_front.push_back(to_neuron);

                } //BOOST_FOREACH( Link )

            } //rawFront iteration ( Neuron )

            //new rawFront formed
            format_front(raw_front);

        } while( raw_front.size() > 0 );

        //We rebuilded cache
        cache.fixed();

    } //regenerateCache

    void
    Net::thread_base(RunDirection direction, RunnerPtr runner, Net *net, unsigned cur_thread, boost::barrier *barrier)
    {
        const NetCache& cache = net->get_cache();

        unsigned layer;
        (direction == ForwardRun) ?  (layer = 0) : (layer = cache.layers.size() - 1);

        const unsigned threads = net->get_work_threads_count();

        do {
            //Process current layer
            for(unsigned i = cur_thread; i < cache.layers[layer].size(); i += threads)
                runner->run( ObjectConstPtr(net), cache.layers[layer][i]); //We pass Net* to runner, because
                                                                    //learning algorithms require
                                                                    //read-only access to Net attributes

            //Wait for other threads
            barrier->wait();
        } while( (direction == ForwardRun && ++layer < cache.layers.size()) || (direction == BackwardRun && layer-- > 0) );
        /*
         * A little comment.
         * Cache structure:
         * Layer1:   thread1_data, thread2_data, ...
         * Layer2:   thread1_data, thread2_data, ...
         * ...
         * LayerN-1: thread1_data, thread2_data, ...
         * LayerN:          <= last layer. ALWAYS empty! (see Net::regenerateCache())
         * cache.size() == N+1;
         * thread_data is vector of Neuron*
         */
    } //thread_base

}; //pann

